crates/ty_python_semantic/src/types.rs

```diff
@@ -7185,6 +7271,158 @@ impl<'db> FieldInstance<'db> {
     }
 }
 
+/// A `typing.NewType` declaration, either from the perspective of the
+/// identity-function-that-acts-like-a-subclass-in-type-expressions returned by the call to
+/// `NewType`, or from the perspective of instances of that subclass. For example:
+///
+/// ```py
+/// from typing import NewType
+/// Foo = NewType("Foo", int)
+/// x = Foo(42)
+/// ```
+///
+/// The revealed types there are:
+/// - `NewType`: `Type::ClassLiteral(ClassLiteral)` with `KnownClass::NewType`.
+/// - `Foo`: `Type::KnownInstance(KnownInstanceType::NewType(NewTypeInstance { .. }))`
+/// - `x`: `Type::NominalInstance(...(NominalInstanceInner::NewType(NewTypeInstance { .. }))`
+#[salsa::interned(debug, heap_size=ruff_memory_usage::heap_size)]
+#[derive(PartialOrd, Ord)]
+pub struct NewTypeInstance<'db> {
+    /// The name of this NewType (e.g. `"Foo"`)
+    #[returns(ref)]
+    pub name: ast::name::Name,
+
+    /// The binding where this NewType is first created.
+    pub definition: Definition<'db>,
+
+    // The base class of this NewType (e.g. `int`), which could be a (specialized) class type or
+    // could be another NewType.
+    pub base: NewTypeBase<'db>,
+}
+
+impl get_size2::GetSize for NewTypeInstance<'_> {}
+
+impl<'db> NewTypeInstance<'db> {
+    // Walk the `NewTypeBase` chain to find the underlying `ClassType`.
+    fn base_class_type(self, db: &'db dyn Db) -> ClassType<'db> {
+        let mut base = self.base(db);
+        loop {
+            match base {
+                NewTypeBase::ClassType(class) => return class,
+                NewTypeBase::NewType(newtype) => base = newtype.base(db),
+            }
+        }
```

Comment:
could shave off a few lines here by making it recursive?

```suggestion
        match self.base(db) {
            NewTypeBase::ClassType(class) => return class,
            NewTypeBase::NewType(newtype) => newtype.base_class_type(db),
        }
```